File: app/src/main/java/com/morsego/app/keyer/MorseDecoder.java
package com.morsego.app.keyer;

import android.os.Handler;
import android.os.Looper;

import com.morsego.app.tree.MorseBinaryTree;

public class MorseDecoder {

    public interface DecoderListener {
        void onPatternChanged(String currentPattern);
        void onTextUpdated(String fullText);
        void onCharacterDecoded(char character);
        default void onTimingFeedback(MorseTiming.PauseEvaluation evaluation) {}
        default void onTimingFailure(MorseTiming.PauseEvaluation evaluation) {}
    }

    private final KeyerSettings settings;
    private DecoderListener listener;
    private final Handler handler;

    private final StringBuilder currentPattern = new StringBuilder();
    private final StringBuilder decodedText = new StringBuilder();

    private Runnable charPauseRunnable;
    private Runnable wordPauseRunnable;
    private Runnable maxLetterPauseRunnable;

    // Timing tracking
    private long lastToneStopTime = 0;
    private long lastCharToneStopTime = 0;

    // Manual / Straight-key timing
    private long toneStartTime = 0;
    private boolean isToneActive = false;

    public MorseDecoder(KeyerSettings settings) {
        this.settings = settings;
        Handler h = null;
        try {
            Looper looper = Looper.getMainLooper();
            if (looper != null) {
                h = new Handler(looper);
            }
        } catch (Exception ignored) {}
        this.handler = h;
    }

    private void postToHandler(Runnable r) {
        if (handler != null) {
            handler.post(r);
        } else {
            r.run();
        }
    }

    private void postDelayedToHandler(Runnable r, long delay) {
        if (handler != null) {
            handler.postDelayed(r, delay);
        }
    }

    private void removeCallbacksFromHandler(Runnable r) {
        if (handler != null && r != null) {
            handler.removeCallbacks(r);
        }
    }

    public void setListener(DecoderListener listener) {
        this.listener = listener;
    }

    public synchronized void onToneStarted() {
        cancelMaxLetterPauseWatcher();
        long now = System.currentTimeMillis();

        if (lastToneStopTime > 0) {
            long pause = now - lastToneStopTime;

            if (currentPattern.length() > 0) {
                // Intra-element pause between dits/dahs of the same character
                MorseTiming.PauseEvaluation eval = MorseTiming.evaluateIntraElementPause(pause, settings.getWpm());
                notifyTimingFeedback(eval);
                if (eval.isTimingFailure) {
                    notifyTimingFailure(eval);
                }
            } else if (lastCharToneStopTime > 0) {
                // Inter-letter pause between letters of a word
                MorseTiming.PauseEvaluation eval = MorseTiming.evaluateLetterPause(pause, settings.getWpm());
                notifyTimingFeedback(eval);
                if (eval.isTimingFailure) {
                    notifyTimingFailure(eval);
                }
            }
        }
    }

    public synchronized void onToneStopped() {
        lastToneStopTime = System.currentTimeMillis();
        restartPauseWatchers();
    }

    public synchronized void onElementReceived(char element) {
        currentPattern.append(element);
        notifyPatternChanged();
        restartPauseWatchers();
    }

    public synchronized void onManualToneState(boolean toneOn) {
        long now = System.currentTimeMillis();
        if (toneOn && !isToneActive) {
            toneStartTime = now;
            isToneActive = true;
            onToneStarted();
            cancelPauseWatchers();
        } else if (!toneOn && isToneActive) {
            isToneActive = false;
            onToneStopped();
            long duration = now - toneStartTime;
            long ditDuration = MorseTiming.ditDurationMs(settings.getWpm());

            char element = (duration >= ditDuration * 2) ? '-' : '.';
            currentPattern.append(element);
            notifyPatternChanged();
            restartPauseWatchers();
        }
    }

    private void restartPauseWatchers() {
        cancelPauseWatchers();

        int wpm = settings.getWpm();
        long charPause = Math.max((long) (MorseTiming.interCharSpaceMs(wpm) * 1.6f), 450L);
        long wordPause = Math.max(MorseTiming.wordSpaceMs(wpm), charPause + 300L);

        charPauseRunnable = () -> {
            commitCharacter();
        };
        postDelayedToHandler(charPauseRunnable, charPause);

        wordPauseRunnable = () -> {
            if (decodedText.length() > 0 && decodedText.charAt(decodedText.length() - 1) != ' ') {
                decodedText.append(' ');
                notifyTextUpdated();
            }
        };
        postDelayedToHandler(wordPauseRunnable, wordPause);
    }

    private void cancelPauseWatchers() {
        if (charPauseRunnable != null) {
            removeCallbacksFromHandler(charPauseRunnable);
            charPauseRunnable = null;
        }
        if (wordPauseRunnable != null) {
            removeCallbacksFromHandler(wordPauseRunnable);
            wordPauseRunnable = null;
        }
    }

    private void scheduleMaxLetterPauseWatcher() {
        cancelMaxLetterPauseWatcher();
        int wpm = settings.getWpm();
        // High generous tolerance: 5.0x inter-character pause with a minimum floor of 2200ms
        long maxWait = Math.max((long) (MorseTiming.interCharSpaceMs(wpm) * 5.0f) + 200L, 2200L);
        maxLetterPauseRunnable = () -> {
            if (listener != null && decodedText.length() > 0) {
                boolean isPt = java.util.Locale.getDefault().getLanguage().equalsIgnoreCase("pt");
                String feedback = isPt ?
                        "Falha: Pausa excessiva entre letras (> máx 5.0x)" :
                        "Failure: Excessive pause between letters (> max 5.0x)";
                MorseTiming.PauseEvaluation eval = new MorseTiming.PauseEvaluation(
                        false, true, feedback, 5.1f);
                notifyTimingFeedback(eval);
                notifyTimingFailure(eval);
            }
        };
        postDelayedToHandler(maxLetterPauseRunnable, maxWait);
    }

    private void cancelMaxLetterPauseWatcher() {
        if (maxLetterPauseRunnable != null) {
            removeCallbacksFromHandler(maxLetterPauseRunnable);
            maxLetterPauseRunnable = null;
        }
    }

    public synchronized void commitCharacter() {
        if (currentPattern.length() == 0) return;

        lastCharToneStopTime = lastToneStopTime > 0 ? lastToneStopTime : System.currentTimeMillis();

        String pattern = currentPattern.toString();
        currentPattern.setLength(0);
        notifyPatternChanged();

        String found = MorseBinaryTree.getInstance().getChar(pattern);
        char c = (found != null && !found.isEmpty()) ? found.charAt(0) : '?';

        decodedText.append(c);
        notifyTextUpdated();

        if (listener != null) {
            listener.onCharacterDecoded(c);
        }

        scheduleMaxLetterPauseWatcher();
    }

    private void notifyTimingFeedback(MorseTiming.PauseEvaluation eval) {
        if (listener != null) {
            postToHandler(() -> listener.onTimingFeedback(eval));
        }
    }

    private void notifyTimingFailure(MorseTiming.PauseEvaluation eval) {
        if (listener != null) {
            postToHandler(() -> listener.onTimingFailure(eval));
        }
    }

    public synchronized void clear() {
        cancelPauseWatchers();
        cancelMaxLetterPauseWatcher();
        lastToneStopTime = 0;
        lastCharToneStopTime = 0;
        currentPattern.setLength(0);
        decodedText.setLength(0);
        notifyPatternChanged();
        notifyTextUpdated();
    }

    public synchronized void backspace() {
        if (decodedText.length() > 0) {
            decodedText.deleteCharAt(decodedText.length() - 1);
            notifyTextUpdated();
        }
    }

    public synchronized String getCurrentPattern() {
        return currentPattern.toString();
    }

    public synchronized String getDecodedText() {
        return decodedText.toString();
    }

    private void notifyPatternChanged() {
        if (listener != null) {
            final String p = currentPattern.toString();
            postToHandler(() -> listener.onPatternChanged(p));
        }
    }

    private void notifyTextUpdated() {
        if (listener != null) {
            final String t = decodedText.toString();
            postToHandler(() -> listener.onTextUpdated(t));
        }
    }
}

File: app/src/main/java/com/morsego/app/keyer/MorseDecoder.kt
package com.morsego.app.keyer

import com.morsego.app.model.MorseDictionary
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch

/**
 * Decodes Morse dit/dah elements and timing into readable text in real-time.
 */
class MorseDecoder(
    private val scope: CoroutineScope = CoroutineScope(Dispatchers.Default),
    private var wpm: Int = 15,
    private val onCharDecoded: (Char) -> Unit = {}
) {
    private val _currentPattern = MutableStateFlow("")
    val currentPattern: StateFlow<String> = _currentPattern.asStateFlow()

    private val _decodedText = MutableStateFlow("")
    val decodedText: StateFlow<String> = _decodedText.asStateFlow()

    private var pauseWatcherJob: Job? = null
    private var lastElementTime = 0L

    // For manual / straight-key timing
    private var toneStartTime = 0L
    private var isToneActive = false

    fun setWpm(newWpm: Int) {
        wpm = newWpm.coerceIn(5, 50)
    }

    /**
     * Called when an element ('.' or '-') is emitted by the Iambic Engine or Keyer
     */
    fun onElementReceived(element: Char) {
        _currentPattern.value += element
        lastElementTime = System.currentTimeMillis()
        restartPauseWatcher()
    }

    /**
     * Called when tone turns on/off in straight key mode
     */
    fun onManualToneState(toneOn: Boolean) {
        val now = System.currentTimeMillis()
        if (toneOn && !isToneActive) {
            toneStartTime = now
            isToneActive = true
            pauseWatcherJob?.cancel()
        } else if (!toneOn && isToneActive) {
            isToneActive = false
            val duration = now - toneStartTime
            val ditDuration = MorseTiming.ditDurationMs(wpm)

            // If duration is longer than 2 dits, classify as dash, else dot
            val element = if (duration >= (ditDuration * 2.0)) '-' else '.'
            onElementReceived(element)
        }
    }

    private fun restartPauseWatcher() {
        pauseWatcherJob?.cancel()
        pauseWatcherJob = scope.launch {
            val charPause = MorseTiming.interCharSpaceMs(wpm)
            val wordPause = MorseTiming.wordSpaceMs(wpm)

            delay(charPause)
            if (isActive && _currentPattern.value.isNotEmpty()) {
                commitCharacter()
            }

            delay(wordPause - charPause)
            if (isActive && _decodedText.value.isNotEmpty() && !_decodedText.value.endsWith(" ")) {
                _decodedText.value += " "
            }
        }
    }

    private fun commitCharacter() {
        val pattern = _currentPattern.value
        _currentPattern.value = ""
        if (pattern.isEmpty()) return

        val char = MorseDictionary.getChar(pattern) ?: '?'
        _decodedText.value += char
        onCharDecoded(char[0])
    }

    fun clear() {
        pauseWatcherJob?.cancel()
        _currentPattern.value = ""
        _decodedText.value = ""
    }

    fun backspace() {
        if (_decodedText.value.isNotEmpty()) {
            _decodedText.value = _decodedText.value.dropLast(1)
        }
    }
}
